kubelet-to-gcm/monitor/kubelet/translate.go

```diff
@@ -304,114 +411,226 @@ func translateCPU(cpu *stats.CPUStats, tsFactory *timeSeriesFactory, startTime t
+	legacyFSMetrics := map[string]*metricMetadata{
+		"disk_total": legacyDiskTotalMD,
+		"disk_used": legacyDiskUsedMD,
+	}
+	return translateFS(volume, fs, tsFactory, startTime, legacyFSMetrics)
+}
+
+func nodeTranslateFS(volume string, fs *stats.FsStats, tsFactory *timeSeriesFactory, startTime time.Time) ([]*v3.TimeSeries, error) {
+	nodeFSMetrics := map[string]*metricMetadata{
+		"disk_total": nodeEphemeralstorageUsedMD,
+		"disk_used": nodeEphemeralstorageTotalMD,
+	}
+	return translateFS(volume, fs, tsFactory, startTime, nodeFSMetrics)
+}
+
+func containerTranslateFS(volume string, rootfs *stats.FsStats, logs *stats.FsStats, tsFactory *timeSeriesFactory, startTime time.Time) ([]*v3.TimeSeries, error) {
+	containerFSMetrics := map[string]*metricMetadata{
+		"disk_used": containerEphemeralstorageUsedMD,
+	}
+
+	var combinedUsage uint64
+	if rootfs != nil {
+		combinedUsage = *rootfs.UsedBytes
+	}
+	if logs != nil {
+		combinedUsage = combinedUsage + *logs.UsedBytes
+	}
+	
+	combinedStats := &stats.FsStats{
+		UsedBytes: &combinedUsage,
+	}
+	if rootfs == nil && logs == nil {
+		combinedStats = nil
+	}
+	return translateFS(volume, combinedStats, tsFactory, startTime, containerFSMetrics)
+}
+
 // translateFS creates all the TimeSeries for a given FsStats and volume name.
-func translateFS(volume string, fs *stats.FsStats, tsFactory *timeSeriesFactory, startTime time.Time) ([]*v3.TimeSeries, error) {
+func translateFS(volume string, fs *stats.FsStats, tsFactory *timeSeriesFactory, startTime time.Time, metrics map[string]*metricMetadata) ([]*v3.TimeSeries, error) {
 	var timeSeries []*v3.TimeSeries
 
-	// First, check that we've been given all the data we need.
 	if fs == nil {
 		return nil, fmt.Errorf("File-system information missing.")
 	}
-	if fs.CapacityBytes == nil {
-		return nil, fmt.Errorf("CapacityBytes is missing from FsStats %v", fs)
-	}
-	if fs.UsedBytes == nil {
-		return nil, fmt.Errorf("UsedBytes is missing from FsStats %v", fs)
-	}
 
 	// For some reason the Kubelet doesn't return when this sample is from,
 	// so we'll use now.
 	now := time.Now()
 
 	resourceLabels := map[string]string{"device_name": volume}
-	// Total disk available.
-	diskTotalPoint := tsFactory.newPoint(&v3.TypedValue{
-		Int64Value:      monitor.Int64Ptr(int64(*fs.CapacityBytes)),
-		ForceSendFields: []string{"Int64Value"},
-	}, startTime, now, diskTotalMD.MetricKind)
-	timeSeries = append(timeSeries, tsFactory.newTimeSeries(resourceLabels, diskTotalMD, diskTotalPoint))
-
-	// Total disk used.
-	diskUsedPoint := tsFactory.newPoint(&v3.TypedValue{
-		Int64Value:      monitor.Int64Ptr(int64(*fs.UsedBytes)),
-		ForceSendFields: []string{"Int64Value"},
-	}, startTime, now, diskUsedMD.MetricKind)
-	timeSeries = append(timeSeries, tsFactory.newTimeSeries(resourceLabels, diskUsedMD, diskUsedPoint))
+	if diskTotalMD, found := metrics["disk_total"]; found {
```

Comment:
I don't like these maps with hardcoded keys here. This seems quite error-prone. Can pass the parameters explicitly?

Reply:
Done.